### DISCUSSION TITLE
Expose SerializeObject type

### DISCUSSION THREAD

**Discussion by andrelandgraf** [remix-run/remix#4270] [Q&A] [answered]
It would be great if Remix could export `SerializeObject` so that I can wrap my data types (e.g., from Prisma) by `SerializeObject` to be compatible with the data exposed by `useLoaderData`.

I love the addition of `useLoaderData<typeof loader>();` to infer the type of the loader function's return value. However, now the return value type will be wrapped by `SerializeObject`.

I understand why `SerializeObject` is required. The problem is that all my custom components and helper functions expect types not wrapped by `SerializeObject`. I am okay with having some functions accept `SerializeObject<User>`, but for that, I would need it to be exposed by Remix! 99% of the time, I don't care about the `createdAt` type if it's a string or date and I would like to avoid deserialization when not necessary.
 
Something like this works if the `loader` function is in scope:

```ts
type LayoutProps = ReturnType<typeof useLoaderData<typeof loader>> & {
  children: React.ReactNode;
};
```

But I believe it would be better if Remix could expose the wrapper types altogether.

**Comment by kiliman** [comment]
Have you checked out `remix-typedjson`?

https://github.com/kiliman/remix-typedjson

**Reply by andrelandgraf** [reply]
Hey Michael! Yes, it's awesome! It's definitely worth it for bigger projects, but in this case, it felt like overkill. `createdAt` and `updatedAt` is all I got to deserialize. It seems like such a common case, that I would expect some kind of out-of-the-box support - at least typing support. But maybe, I should just type with `any` then. :D

**Comment by gnapse** [comment]
I'm starting off with remix, and hit this wall fairly quickly. As soon as you want to extract a piece of your markup from a route component into a new component, the new component often need objects from the loader data to be passed as props. But how can you declare the types of those props?

Anyway, hoping this discussion will lead to something that solves this need.

**Reply by gnapse** [reply]
In the short-term I was able to solve it in this way for now:

> **Context**
> I extracted the markup that renders a note into a component `NoteCard` that needs a note object that comes from loader data

```typescript
type SerializedObject<T extends object> = {
  [k in keyof T]: T[k] extends Date ? string : T[k];
};

type NoteCardProps = {
  note: SerializedObject<Note>;
};
```

The rationale is that the main hurdle I was facing was with the `createdAt` field not being recognized as a string inside `NoteCard` if I declared the `note` as `note: Note`. When the markup is used directly inside the route component, the `note` object is not of type `note` but of type `note: SerializeObject<UndefinedToOptional<Note>>`. However, these wrapping generic types are not exposed by Remix, hence I cannot use it in my code.

By creating my own wrapper that converts date-typed fields into string-typed, it works both on this side, and it seems to accept the note from the loader as well, no TS errors. At least for date fields. Not sure if this solution is generic enough for all cases.

**Comment by andrelandgraf** [comment, marked as answer]
@gnapse TIL Remix exposes the `SerializeFrom` type (e.g.,`import type { SerializeFrom } from '@remix-run/node';`). You can use that type to wrap your custom types!

```ts
import type { SerializeFrom } from '@remix-run/node';

type NoteCardProps = {
  note: SerializeFrom<Note>;
};
```

Thanks @sergiodxa for telling me about `SerializeFrom` 🙏
